src/realm/error_codes.yml

```diff
@@ -0,0 +1,48 @@
+error_categories:
+  - generic_error
+  - logic_error
+  - runtime_error
+  - invalid_argument
```

Comment:
These categories are a bit confusing. `invalid_argument` and `logic_error` kinda make sense, but what is the difference between `generic_error` and `runtime_error`?. To me, that seems like a "We don't know what this is"-category and could just be combined into `runtime_error` IMO.

Going from https://github.com/realm/realm-core/issues/4852. At least the Kotlin SDK would like these 3 top-level categories:

- `IllegalArgument`, i.e. user provided an argument that wasn't allowed. This maps to `invalid_argument`.
- `IllegalState`, i.e. an operation was performed on an object which was not in the correct state. This maps to `logic_error`
- `RealmError`, i.e. Core hit an unexpected bug. In some cases this might be an assertion. But this seems to map to just `runtime_error`.

Reply:
I don't know if this setup supports nested categories, but I suspect that could be useful as well, e.g. so `runtime_error` could have categories like `query`, `memory_mapping`, `filesystem` etc.

Reply:
The `generic_error` is only used in some catch all code. So we will not actively throw exceptions in this category.  We only have one kind of logic error. I think we only have a few that can be seen as "core internal" exceptions - we normally use asserts for these checks.
Runtime errors are in a separate category. These are errors that happen due to external conditions that cannot be predicted.

The intention is to support nested categories.